### PR TITLE
Store and forward

### PR DESCRIPTION
**Pull request by moxie0**

### CODE REVIEW

File: libtextsecure/src/main/java/org/whispersystems/textsecure/internal/push/PushServiceSocket.java [hunk 1]
```diff
@@ -539,35 +533,21 @@ private HttpURLConnection getConnection(String urlFragment, String method, Strin
...
-
-      return BlacklistingTrustManager.createFor(trustManagerFactory.getTrustManagers());
-    } catch (KeyStoreException | CertificateException | NoSuchAlgorithmException | IOException kse) {
-      throw new AssertionError(kse);
-    }
-  }
-
   private static class GcmRegistrationId {
     private String gcmRegistrationId;
+    private boolean webSocketChannel;
```
Comment (mcginty) [inline libtextsecure/src/main/java/org/whispersystems/textsecure/internal/push/PushServiceSocket.java]: this doesn't look like it's used anywhere

File: libtextsecure/src/main/java/org/whispersystems/textsecure/api/TextSecureMessagePipe.java [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+      }
+    }
+  }
+
+  public void shutdown() throws IOException {
+    websocket.disconnect();
+  }
+
+  private boolean isTextSecureEnvelope(WebSocketRequestMessage message) {
+    return "PUT".equals(message.getVerb()) && "/api/v1/message".equals(message.getPath());
```
Comment (mcginty) [inline libtextsecure/src/main/java/org/whispersystems/textsecure/api/TextSecureMessagePipe.java]: should the endpoint refer to the PushServiceSocket constants for when the API changes?

File: libtextsecure/protobuf/WebSocketResources.proto [hunk 3]
```diff
@@ -0,0 +1,46 @@
+/**
+ * Copyright (C) 2014 Open WhisperSystems
```
Comment (mcginty) [inline libtextsecure/protobuf/WebSocketResources.proto]: 2015

File: libtextsecure/src/main/java/org/whispersystems/textsecure/internal/websocket/WebSocketConnection.java [hunk 4]
```diff
@@ -0,0 +1,292 @@
...
+
+    client.sendMessage(message.toByteArray());
+  }
+
+  private synchronized void sendKeepAlive() throws IOException {
+    if (keepAliveSender != null) {
+      client.sendMessage(WebSocketMessage.newBuilder()
+                                         .setType(WebSocketMessage.Type.REQUEST)
+                                         .setRequest(WebSocketRequestMessage.newBuilder()
+                                         .setId(System.currentTimeMillis())
```
Comment (mcginty) [inline libtextsecure/src/main/java/org/whispersystems/textsecure/internal/websocket/WebSocketConnection.java]: identation

File: src/org/thoughtcrime/securesms/PassphraseRequiredMixin.java [hunk 5]
```diff
@@ -27,11 +28,13 @@
     initializeNewKeyReceiver(activity);
     initializeFromMasterSecret(activity);
     KeyCachingService.registerPassphraseActivityStarted(activity);
+    MessageRetrievalService.registerActivityStarted(activity);
   }
 
   public <T extends Activity & PassphraseRequiredActivity> void onPause(T activity) {
     removeNewKeyReceiver(activity);
     KeyCachingService.registerPassphraseActivityStopped(activity);
+    MessageRetrievalService.registerActivityStopped(activity);
```
Comment (mcginty) [inline src/org/thoughtcrime/securesms/PassphraseRequiredMixin.java]: is there a reason we don't do these on `onStop()`? In `onPause` the Activities could still be visible, just partially obscured by a dialog or something.

File: src/org/thoughtcrime/securesms/service/RegistrationService.java [hunk 6]
```diff
@@ -245,9 +245,11 @@ private void handleCommonRegistration(MasterSecret masterSecret, TextSecureAccou
       setState(new RegistrationState(RegistrationState.STATE_GCM_REGISTERING, number));
 
       String gcmRegistrationId = GoogleCloudMessaging.getInstance(this).register(GcmRefreshJob.REGISTRATION_ID);
-      TextSecurePreferences.setGcmRegistrationId(this, gcmRegistrationId);
       accountManager.setGcmId(Optional.of(gcmRegistrationId));
 
+      TextSecurePreferences.setGcmRegistrationId(this, gcmRegistrationId);
+      TextSecurePreferences.setWebsocketRegistered(this, true);
```
Comment (mcginty) [inline src/org/thoughtcrime/securesms/service/RegistrationService.java]: One must re-register to enable websockets? Looks like thisis the only place setWebsocketRegistered is called.

File: src/org/thoughtcrime/securesms/service/MessageRetrievalService.java [hunk 7]
```diff
@@ -0,0 +1,180 @@
...
+  }
+
+  @Override
+  public void run() {
+    while (true) {
+      Log.w(TAG, "Waiting for websocket state change....");
+      waitForConnectionNecessary();
+
+      Log.w(TAG, "Making websocket connection....");
+      TextSecureMessagePipe pipe = receiver.createMessagePipe();
```
Comment (mcginty) [inline src/org/thoughtcrime/securesms/service/MessageRetrievalService.java]: seems like this call would also want to wait on network availability and whether we're registered for websockets, as it looks like it internally connects and does a linear backoff until somebody calls `disconnect()`

File: src/org/thoughtcrime/securesms/service/MessageRetrievalService.java [hunk 8]
```diff
@@ -0,0 +1,180 @@
...
+    else if (ACTION_PUSH_RECEIVED.equals(intent.getAction()))     incrementPushReceived();
+
+    return START_STICKY;
+  }
+
+  @Override
+  public void run() {
+    while (true) {
+      Log.w(TAG, "Waiting for websocket state change....");
+      waitForConnectionNecessary();
```
Comment (mcginty) [inline src/org/thoughtcrime/securesms/service/MessageRetrievalService.java]: this will just be an infinitely looping if isWebsocketRegistered is false?

File: libtextsecure/src/main/java/org/whispersystems/textsecure/internal/push/PushServiceSocket.java (hunk 1 above)
Comment (moxie0) [inline libtextsecure/src/main/java/org/whispersystems/textsecure/internal/push/PushServiceSocket.java]: This is a JSON POJO. It's transmitted to the server.

File: libtextsecure/src/main/java/org/whispersystems/textsecure/api/TextSecureMessagePipe.java (hunk 2 above)
Comment (moxie0) [inline libtextsecure/src/main/java/org/whispersystems/textsecure/api/TextSecureMessagePipe.java]: This endpoint is unrelated to PushServiceSocket, it's strictly a websocket thing.

File: src/org/thoughtcrime/securesms/PassphraseRequiredMixin.java (hunk 5 above)
Comment (moxie0) [inline src/org/thoughtcrime/securesms/PassphraseRequiredMixin.java]: I think they used to be onStop, but that wasn't called when the screen was off on some devices.

File: src/org/thoughtcrime/securesms/service/RegistrationService.java (hunk 6 above)
Comment (moxie0) [inline src/org/thoughtcrime/securesms/service/RegistrationService.java]: Yes, it's an incremental rollout.  I want to start with new users rather than doing it all at once.

File: src/org/thoughtcrime/securesms/service/MessageRetrievalService.java (hunk 8 above)
Comment (moxie0) [inline src/org/thoughtcrime/securesms/service/MessageRetrievalService.java]: Yep, idle thread.

File: libtextsecure/src/main/java/org/whispersystems/textsecure/internal/websocket/WebSocketConnection.java (hunk 4 above)
Comment (moxie0) [inline libtextsecure/src/main/java/org/whispersystems/textsecure/internal/websocket/WebSocketConnection.java]: ?

Comment (mcginty) [inline libtextsecure/src/main/java/org/whispersystems/textsecure/internal/websocket/WebSocketConnection.java]: *indentation - the chained methods from `WebSocketRequestMessage.newBuilder()` are aligned with those of `WebSocketMessage.newBuilder()`, making it kind of confusing to read.

File: src/org/thoughtcrime/securesms/service/RegistrationService.java (hunk 6 above)
Comment (mcginty) [inline src/org/thoughtcrime/securesms/service/RegistrationService.java]: got it, makes sense

File: src/org/thoughtcrime/securesms/service/MessageRetrievalService.java (hunk 7 above)
Comment (moxie0) [inline src/org/thoughtcrime/securesms/service/MessageRetrievalService.java]: If the network is unavailable, that disconnect call will happen a maximum of 1min later.

**Comment by moxie0** [comment]
updated the copyright header and indentation

File: libtextsecure/src/main/java/org/whispersystems/textsecure/api/TextSecureMessagePipe.java [hunk 9]
```diff
@@ -0,0 +1,87 @@
...
+  public TextSecureEnvelope read(long timeout, TimeUnit unit, MessagePipeCallback callback)
+      throws TimeoutException, IOException, InvalidVersionException
+  {
+    while (true) {
+      WebSocketRequestMessage  request  = websocket.readRequest(unit.toMillis(timeout));
+      WebSocketResponseMessage response = createWebSocketResponse(request);
+
+      try {
+        if (isTextSecureEnvelope(request)) {
+          TextSecureEnvelope envelope = new TextSecureEnvelope(request.getBody().toByteArray(),
```
Comment (mcginty) [inline libtextsecure/src/main/java/org/whispersystems/textsecure/api/TextSecureMessagePipe.java]: since body is optional in the protobuf, could someone just send an envelope without one and cause NPE here?

File: libtextsecure/src/main/java/org/whispersystems/textsecure/internal/push/PushServiceSocket.java (hunk 1 above)
Comment (mcginty) [inline libtextsecure/src/main/java/org/whispersystems/textsecure/internal/push/PushServiceSocket.java]: why not do `private boolean webSocketChannel = true;`?

File: libtextsecure/src/main/java/org/whispersystems/textsecure/internal/websocket/WebSocketConnection.java [hunk 10]
```diff
@@ -0,0 +1,292 @@
...
+                                         .toByteArray());
+    }
+  }
+
+  private synchronized void onMessage(byte[] payload) {
+    Log.w(TAG, "WSC onMessage()");
+    try {
+      WebSocketMessage message = WebSocketMessage.parseFrom(payload);
+
+      Log.w(TAG, "Message Type: " + message.getType().getNumber());
```
Comment (mcginty) [inline libtextsecure/src/main/java/org/whispersystems/textsecure/internal/websocket/WebSocketConnection.java]: same, can Type be null?

File: libtextsecure/src/main/java/org/whispersystems/textsecure/api/TextSecureMessagePipe.java (hunk 9 above)
Comment (moxie0) [inline libtextsecure/src/main/java/org/whispersystems/textsecure/api/TextSecureMessagePipe.java]: "someone" here is the server, which i'm willing to trust with an npe

File: libtextsecure/src/main/java/org/whispersystems/textsecure/internal/push/PushServiceSocket.java (hunk 1 above)
Comment (moxie0) [inline libtextsecure/src/main/java/org/whispersystems/textsecure/internal/push/PushServiceSocket.java]: not really how json pojos are written, and i want to be explicit to the caller

File: libtextsecure/src/main/java/org/whispersystems/textsecure/internal/websocket/WebSocketConnection.java [hunk 11]
```diff
@@ -0,0 +1,292 @@
...
+    if (keepAliveSender != null) {
+      keepAliveSender.shutdown();
+      keepAliveSender = null;
+    }
+
+    notifyAll();
+  }
+
+  private synchronized void onConnected() {
+    if (client != null) {
```
Comment (mcginty) [inline libtextsecure/src/main/java/org/whispersystems/textsecure/internal/websocket/WebSocketConnection.java]: `if (keepAliveSender != null)`?

File: libtextsecure/src/main/java/org/whispersystems/textsecure/internal/websocket/WebSocketConnection.java (hunk 10 above)
Comment (moxie0) [inline libtextsecure/src/main/java/org/whispersystems/textsecure/internal/websocket/WebSocketConnection.java]: also set by the server, not the sender

File: src/org/thoughtcrime/securesms/jobs/PushDecryptJob.java [hunk 12]
```diff
@@ -221,10 +221,12 @@ private void handleLegacyMessage(MasterSecret masterSecret, TextSecureEnvelope e
...
   private void handleDuplicateMessage(MasterSecret masterSecret, TextSecureEnvelope envelope) {
-    Pair<Long, Long> messageAndThreadId = insertPlaceholder(masterSecret, envelope);
-    DatabaseFactory.getEncryptingSmsDatabase(context).markAsDecryptDuplicate(messageAndThreadId.first);
+    // Let's start ignoring these now.
 
-    MessageNotifier.updateNotification(context, masterSecret, messageAndThreadId.second);
+//    Pair<Long, Long> messageAndThreadId = insertPlaceholder(masterSecret, envelope);
+//    DatabaseFactory.getEncryptingSmsDatabase(context).markAsDecryptDuplicate(messageAndThreadId.first);
+//
+//    MessageNotifier.updateNotification(context, masterSecret, messageAndThreadId.second);
```
Comment (mcginty) [inline src/org/thoughtcrime/securesms/jobs/PushDecryptJob.java]: commented code probably unnecessary

**Comment by mcginty** [comment]
haven't run it yet, but code lgtm :)

File: libtextsecure/src/main/java/org/whispersystems/textsecure/internal/websocket/WebSocketConnection.java (hunk 11 above)
Comment (moxie0) [inline libtextsecure/src/main/java/org/whispersystems/textsecure/internal/websocket/WebSocketConnection.java]: Not sure what you mean, we need to construct a new keepaliversender for every connect event.  We only check the 'client' status to determine if a close call happened before this callback completed.

Comment (mcginty) [inline libtextsecure/src/main/java/org/whispersystems/textsecure/internal/websocket/WebSocketConnection.java]: Right, it was just not easy to tell that the previous Thread referenced by keepalivesender was shutdown before we replaced the reference.
